### PR TITLE
Fixes #35626 - Registration & proxy's registration_url

### PR DESCRIPTION
**Pull request by stejskalleos**
Use registration_url from Registration SP module
for registration command.
    
RFC: https://community.theforeman.org/t/rfc-host-registration-and-load-balancers/30462
Katello: https://projects.theforeman.org/issues/35627
Smart Proxy: https://projects.theforeman.org/issues/35639
Installer: TODO

### CODE REVIEW

File: app/controllers/concerns/foreman/controller/registration_commands.rb [hunk 1]
```diff
@@ -22,7 +22,12 @@ def insecure
   def endpoint
     return global_registration_url if registration_params['smart_proxy_id'].blank?
 
-    "#{@smart_proxy.url}/register"
+    template_settings = @smart_proxy.smart_proxy_features
+                                    .find_by(feature: Feature.find_by(name: 'Templates'))
+                                    .settings
+    url = template_settings['template_url'].presence || @smart_proxy.url
```
Comment (ekohl) [inline app/controllers/concerns/foreman/controller/registration_commands.rb]: There's a much shorter way for this:
```suggestion
    url = @smart_proxy.setting('Templates', 'template_url') || @smart_proxy.url
```

File: app/controllers/registration_commands_controller.rb [hunk 2]
```diff
@@ -52,7 +52,14 @@ def host_groups_json
   end
 
   def smart_proxies
-    SmartProxy.with_features('Templates') & SmartProxy.with_features('Registration')
+    proxies = SmartProxy.with_features('Templates') & SmartProxy.with_features('Registration')
+    proxies.map do |proxy|
+      url = proxy.smart_proxy_features
+                 .find_by(feature: Feature.find_by(name: 'Templates'))
+                 .settings['template_url']
+                 .presence || proxy.url
```
Comment (ekohl) [inline app/controllers/registration_commands_controller.rb]:
```suggestion
      url = proxy.setting('Templates', 'template_url') || proxy.url
```

**Review by ekohl — COMMENTED** [review]
My biggest concern here is that the template URL is HTTP and not HTTPS. Normally registration uses HTTPS, right? Is that intended?

**Comment by stejskalleos** [comment]
> My biggest concern here is that the template URL is HTTP and not HTTPS. Normally registration uses HTTPS, right? Is that intended?

I wouldn't say it's intended, but if users use `template_url` for haproxy (as [described in this comment](https://community.theforeman.org/t/rfc-host-registration-and-load-balancers/30462/18)) it can be https URL.

**Comment by ekohl** [comment]
It indeed can be HTTPS, but in that case (AFAIK) you can't really kickstart since we don't set up certificates. At least back in the day Anaconda couldn't even retrieve its kickstart over HTTPS. Perhaps that's better now, but that's why it HTTP and not HTTPS. That's why I'd still suggest to introduce a new setting on the registration module. That way users don't have to choose between the two.

**Comment by ehelms** [comment]
> That's why I'd still suggest to introduce a new setting on the registration module. That way users don't have to choose between the two.

Expanding on this, the suggestion is to add a new setting into registration module (https://github.com/theforeman/smart-proxy/blob/develop/config/settings.d/registration.yml.example) that is named `registration_url` and be set directly at installation time and is reported back to Foreman and keeps it separate from the templates URL?

Does the registration workflow hit directly the templates URL?

**Comment by stejskalleos** [comment]
> Does the registration workflow hit directly the templates URL?

Yes, it's used in the end of registration when we are calling back to Foreman to update build status, in [host_init_config_default template](https://github.com/theforeman/foreman/blob/develop/app/views/unattended/provisioning_templates/host_init_config/host_init_config_default.erb#L37), nothing else.

Having `registration_url` for LB case sounds like a good idea, implementation shouldn't be big problem and it would solve the http problem with `templates_url`, so :+1: from me.

**Comment by ekohl** [comment]
> Expanding on this, the suggestion is to add a new setting into registration module (https://github.com/theforeman/smart-proxy/blob/develop/config/settings.d/registration.yml.example) that is named `registration_url` and be set directly at installation time and is reported back to Foreman and keeps it separate from the templates URL?

Exactly.

File: app/controllers/registration_commands_controller.rb [hunk 3]
```diff
@@ -52,7 +52,11 @@ def host_groups_json
   end
 
   def smart_proxies
-    SmartProxy.with_features('Templates') & SmartProxy.with_features('Registration')
+    proxies = SmartProxy.with_features('Templates') & SmartProxy.with_features('Registration')
+    proxies.map do |proxy|
+      url = proxy.setting('Registration', 'registration_url').presence || proxy.url
```
Comment (ekohl) [inline app/controllers/registration_commands_controller.rb]: Some design debate: this already includes the registration commands concern. Can't we introduce a helper method `registration_url(proxy)` in the concern that does this and then reuse it in two places?

I also wonder if this URL should be the full endpoint or if it should be named `base_url`. Reading the JS it's only used in the helper text. I'd think it makes sense to provide the complete URL (including `/register`) to keep things consistent.

**Review by ekohl — COMMENTED** [review]
ACK on the Ruby side, but I haven't done anything with the JS side for a while so I'd appreciate if someone else could take a look.

**Review by Ron-Lavi — APPROVED** [review]
Looks good to me, thanks @stejskalleos 👍🏼

**Comment by Ron-Lavi** [comment]
Should we try merging https://github.com/Katello/katello/pull/10315, https://github.com/theforeman/smart-proxy/pull/850 all at the same time? any other concerns before merging?

**Comment by stejskalleos** [comment]
> Should we try merging https://github.com/Katello/katello/pull/10315, https://github.com/theforeman/smart-proxy/pull/850 all at the same time? any other concerns before merging?

Katello can wait few days, but smart-proxy should go together with this PR

**Pull request merged**